### DISCUSSION TITLE
Named bindings

### DISCUSSION THREAD

**Discussion by murat-mehmet** [mscharley/dot#252] [Q&A] [answered]
In Inversify I was using named bindings for many useful things. Is there an alternative here or any plan to implement it?

**Comment by mscharley** [comment, marked as answer]
There is nothing currently, but [this is the proposal](https://github.com/mscharley/dot/discussions/233) that was supposed to help deal with this sort of situation. In truth, I've never needed named bindings so I'm not 100% across what the needs behind it are though. Is there anything in that proposal which would be missing for your usecase?

**Reply by murat-mehmet** [reply]
I mainly use it for 2 purposes.

1. Giving a tag to the logger services. Check this example:
```js
// bindings
container
  .bind(REQUEST)
  .toDynamicValue(context => context.currentRequest.parentRequest)
  .inTransientScope();
container.bind(LOGGER).to(LoggerService).inTransientScope();

// logger.service.ts
@injectable()
export class LoggerService {
  constructor(@inject(REQUEST) request: interfaces.Request) {
    this.prefix = request.target.getNamedTag()?.value;
  }

  log(text: string) {
    console.log(`[${this.prefix}] ${text}`;
  }
}

// connection.service.ts
@injectable()
export class ConnectionService {
  @inject(LOGGER) @named('CONN') logger: LoggerService;

  // this.logger.log('connected') prints: [CONN] connected
}
```

2. Binding components per type. Example:
```jsx
// bindings
  container
    .bind(TIMELINE_ITEM)
    .toConstantValue(PhotoItem)
    .whenTargetNamed('photo');
  container
    .bind(TIMELINE_ITEM)
    .toConstantValue(VideoItem)
    .whenTargetNamed('video');

// Timeline.tsx
export const Timeline  = () => {
  const container = useContainer();
  const items = [{ type: 'photo' }, { type: 'video' }]; // fetched data
  return items.map(item => {
    const ItemComponent = container.getNamed<FC>(TIMELINE_ITEM, item.type);
    return <ItemComponent />;
  });
}
```

The proposal seems like handles the second case but for the first case we need a way to access the metadata in `toDynamicValue`

**Reply by mscharley** [reply]
Interesting. As you say, the second example should be handled by the proposal as-is. Your first example should be able to be handled pretty simply using `toFactory()` as well, if we provide the metadata object as part of the context provided to the factory.

The one caveat for the first example would be that the context would get a `Partial<Metadata>` because that's what a call to `.get()` would provide. This isn't any worse than your example, but it bears mentioning since type safety is a goal of this library.

The other possible concern is on the binding side. The original intention was that all bindings would have to provide a full metadata object to make filtering to them possible. Maybe that restriction is relaxed for `.toFactory()`. But I suspect this will make caching for non-transient bindings more complicated since we can't just cache using the binding object as a key like we currently do, we would need a separate cache entry for each used set of metadata in the binding.

eg, how to handle the following psuedocode?

```typescript
// bindings
container
  .bind(REQUEST)
  .toDynamicValue(context => context.currentRequest.parentRequest)
  .inTransientScope();
container.bind(LOGGER).inSingletonScope().toFactory(({ metadata }) => new LoggerService(metadata.name));
container.bind(ConnectionService).inTransientScope().toSelf();

// logger.service.ts
export class LoggerService {
  constructor(private private?: string) { }

  log(text: string) {
    console.log(`[${this.prefix}] ${text}`;
  }
}

// connection.service.ts
@injectable()
export class ConnectionService {
  @inject(LOGGER, { name: "CONN" }) logger: LoggerService;

  // this.logger.log('connected') prints: [CONN] connected
}

// These expectations should probably hold true.
const s1 = container.get(ConnectionService);
const s2 = container.get(ConnectionService);
const logger = container.get(LoggerService, { name: "HELLO" });

expect(s1.logger).toBe(s2.logger);
expect(s1.logger).not.toBe(logger);

// But what about these?
const l2 = container.get(LoggerService);
const l3 = container.get(LoggerService);

expect(logger).not.toBe(l2);
expect(l2).toBe(l3);
```

**Reply by murat-mehmet** [reply]
> The one caveat for the first example would be that the context would get a Partial<Metadata> because that's what a call to .get() would provide. This isn't any worse than your example, but it bears mentioning since type safety is a goal of this library.

Yes, Partial is ok here. Also If the user provides no metadata in `.get()`, then the metdata should be an empty object `{}` in `toFactory` and not undefined.

> The other possible concern is on the binding side. The original intention was that all bindings would have to provide a full metadata object to make filtering to them possible. Maybe that restriction is relaxed for .toFactory(). But I suspect this will make caching for non-transient bindings more complicated since we can't just cache using the binding object as a key like we currently do, we would need a separate cache entry for each used set of metadata in the binding.

In this example: 
```js
container.bind(LOGGER).inSingletonScope().toFactory(({ metadata }) => new LoggerService(metadata.name));
```
We clearly want LoggerService to be a singleton independent from metadata. The first call to `.get()` will get it cached with whatever metadata it receives. The Inversify equivalent would be like [this](https://stackblitz.com/edit/inversify-playground-5me53j?file=index.ts). If the user wants separate singletons per dynamic metadata, he can handle it in a transient scoped `toFactory`.

Static metadata, on the other hand, can already be defined as a singleton without any problem as in your proposal:
```js
container.bind(LOGGER).inSingletonScope().withMetadata({ name: "CONN" }).toFactory(({ metadata }) => new LoggerService(metadata.name);
container.bind(LOGGER).inSingletonScope().withMetadata({ name: "HELLO" }).toFactory(({ metadata }) => new LoggerService(metadata.name));
```

> // These expectations should probably hold true.
const s1 = container.get(ConnectionService);
const s2 = container.get(ConnectionService);
const logger = container.get(LoggerService, { name: "HELLO" });

```js
expect(s1.logger).toBe(s2.logger); // true
expect(s1.logger).not.toBe(logger); // FALSE <- it is defined as a singleton without using withMetadata so the logger objects are the same
```

> // But what about these?
const l2 = container.get(LoggerService);
const l3 = container.get(LoggerService);

```js
expect(logger).not.toBe(l2); // false
expect(l2).toBe(l3); // true
```

**Reply by mscharley** [reply]
This isn't to say you're wrong, the inversify example is an interesting datapoint. But I've explicitly tried to distance this project from inversify a little bit and re-examine and understand each feature that makes it into `dot` to make sure I understand it and that everything is internally consistent in how it works. This is one major reason that named bindings don't exist yet - I didn't understand the motivations around why you might want to use it, so I didn't feel qualified to make calls about how it should work. This isn't a fork of inversify, it's a ground-up rebuild. In particular, scopes and caching is one area that bit me several times with Inversify, and this example is actually one area I disagree with them.

Using your alternative example, this is actually what I would expect to happen:

```typescript
container.bind(LOGGER).inSingletonScope().withMetadata({ name: "CONN" }).toFactory(({ metadata }) => new LoggerService(metadata.name);
container.bind(LOGGER).inSingletonScope().withMetadata({ name: "HELLO" }).toFactory(({ metadata }) => new LoggerService(metadata.name));

// These expectations should probably hold true.
const s1 = container.get(ConnectionService);
const s2 = container.get(ConnectionService);
const logger = container.get(LoggerService, { name: "HELLO" });

expect(s1.logger).toBe(s2.logger); // true
expect(s1.logger).not.toBe(logger); // true - s1.logger is the CONN logger, logger is the HELLO logger

const l2 = container.get(LoggerService); // Exception: ambiguous binding found, multiple bindings match this request but a single element was requested.
const loggers = container.get(withOptions(LoggerService, { multiple: true}));
expect(loggers).toBe([s1.logger, logger]); // TRUE - returns an array of all loggers regardless of metadata.
```

This actually speaks to my own use-case for the metadata system, which is listing stuff:

```typescript
container.bind(QUEUE).withMetadata({ name: "Hello", type: "sqs" }).to(HelloQueue);
container.bind(QUEUE).withMetadata({ name: "Goodbye", type: "bull" }).to(GoodbyeQueue);

@injectable(withMetadata(QUEUE, { type: "sqs" }, { multiple: true }))
class SQSManager {
  constructor(private queues: QUEUE[]) {}
}

container.get(QUEUE, { name: "Hello" }).add({ name: "Matt" });
```

ie, allowing for selecting subsets of all the related bindings based on the relevant metadata attached to the bindings.

**Reply by mscharley** [reply]
I actually like this a lot, I think that requiring metadata objects even for factory bindings makes a tonne of sense and eases a lot of the issues.

**Reply by mscharley** [reply]
My concern is that I don't think this meets the spirit of your first example which seems to be allowing the use of arbitrarily named loggers. Only requiring metadata for factory bindings which would be cached might be one solution there I guess. 

One other feature that `dot` has which Inversify doesn't is [container validation](https://mscharley.github.io/dot/docs/dot/interfaces/container/validate/). I don't think this would be impossible, but the number of special cases here could make the validation logic harder to write and maintain. There's already some gaps in the validation algorithm and I'd rather not make that gap wider.

**Reply by murat-mehmet** [reply]
Sorry for the confusion. I think we are talking about the same thing. Let me clarify what I meant with further examples.

The following example matches ur latest assertions.
```js
// Example 1
container.bind(LOGGER).inSingletonScope().withMetadata({ name: "CONN" }).toFactory(({ metadata }) => new LoggerService(metadata.name);
container.bind(LOGGER).inSingletonScope().withMetadata({ name: "HELLO" }).toFactory(({ metadata }) => new LoggerService(metadata.name));

const connLogger = container.get(LoggerService, { name: "CONN" });
const helloLogger = container.get(LoggerService, { name: "HELLO" });

expect(connLogger).not.toBe(helloLogger); // true <- Two different singletons
const l2 = container.get(LoggerService); // Exception: ambiguous binding found, multiple bindings match this request but a single element was requested.
const loggers = container.get(withOptions(LoggerService, { multiple: true}));
expect(loggers).toBe([connLogger, helloLogger]);
```

The following example is the one I intended to present.
```js
// Example 2
// Singleton with NO withMetadata
container.bind(LOGGER).inSingletonScope().toFactory(({ metadata }) => new LoggerService(metadata.name);

const connLogger = container.get(LoggerService, { name: "CONN" });
const helloLogger = container.get(LoggerService, { name: "HELLO" });

expect(connLogger).not.toBe(helloLogger); // FALSE <- One singleton would be cached at first call to .get() 
expect(helloLogger.prefix).toBe('CONN') // TRUE <- because the first call was made with CONN

const l2 = container.get(LoggerService); // Won't throw an exception.
const loggers = container.get(withOptions(LoggerService, { multiple: true})); // This could just return the one singleton.
expect(loggers).toBe([connLogger]);
```
The example 2 is useless of course, this kind of LoggerService makes sense only in transient scope. 

You would cache metadata only when withMetadata is specified in bindings as in Example 1. I think this is also not against your use case, right?

**Reply by mscharley** [reply]
I'm inclined to force `withMetadata()` for all bindings except one outlier - transient factory bindings.

In all situations other than factory bindings, metadata is necessary to distinguish between different bound values, so must be provided. Factorys however are in the unique position of being able to dynamically return a different value based on the metadata provided.

Caching factory bindings only makes sense when you have explicit metadata for the binding. Take for example the following:

```typescript
container.bind(LOGGER).inSingletonScope().toFactory(({ metadata }) => new LoggerService(metadata.name));

@injectable(withMetadata(LOGGER, { name: "A-LOGGER" }))
class A {
  public constructor(private logger: LOGGER) { }
}

@injectable(withMetadata(LOGGER, { name: "B-LOGGER" }))
class B {
  public constructor(private logger: LOGGER) { }
}

@injectable(A, B)
class C {
  public constructor(private a: A, private b: B) { }
}

const c = container.get(C);
c.a.logger.name // What is this value? "A-LOGGER" or "B-LOGGER"?
```

The answer is that this is deterministic in `dot` for any given configuration thanks to the query planner, but it is brittle and will easily change with very small changes to the dependency tree.

I'm inclined to make the following rules around this:

```typescript
// These are allowed:
container.bind(LOGGER).inSingletonScope().withMetadata({ name: "CONN" }).toFactory(({ metadata }) => new LoggerService(metadata.name);
container.bind(LOGGER).inTransientScope().withMetadata({ name: "CONN" }).toFactory(({ metadata }) => new LoggerService(metadata.name);
container.bind(LOGGER).inTransientScope().toFactory(({ metadata }) => new LoggerService(metadata.name));

// These might be allowed in the future, but would rely on smarter caching that can tell metadata input apart rather than caching purely based on the binding. This would mean that A-LOGGER and B-LOGGER above would both be cached separately.
container.bind(LOGGER).inSingletonScope().toFactory(({ metadata }) => new LoggerService(metadata.name));

// These are definitely errors.
container.bind(LOGGER).toConstantValue(new LoggerService("CONN"));
container.bind(LOGGER).inTransientScope().toDynamicValue([], () => new LoggerService("CONN"));
```

This still leaves one outstanding question that I just thought of. What would the following example do?

```typescript
container.bind(LOGGER).inTransientScope().toFactory(({ metadata }) => new LoggerService(metadata.name));

const loggers = container.get(withOptions(LOGGER, { multiple: true, optional: true }, {}));

// Option A:
expect(loggers).toBe([new LoggerService(undefined)]);
// Option B:
expect(loggers).toBe([]);
// Option C:
expect(loggers).toThrow(new Error("Ambiguous binding, the only binding available for a multiple injection was a factory binding"));
```

I think option C might make more sense, this feels like an edge case on an edge case and something that doesn't make a whole lot of sense to attempt. On one hand you've setup the bindings to allow for arbitrary creation of objects based on input, but on the other you're trying to fetch a list of existing objects.

**Reply by murat-mehmet** [reply]
> I'm inclined to force withMetadata() for all bindings except one outlier - transient factory bindings.

That makes sense and covers all the use cases I can think of.

> This still leaves one outstanding question that I just thought of. What would the following example do?

Option A and C both seem ok. Option C looks a bit strict though. You have specified `optional: true` which I would expect it to prevent throwing any exception here. In this case I would expect it to unwrap like option A. Omitting `optional` could throw like in option C.

**Reply by mscharley** [reply]
The `optional` option is really about allowing the situation where there are no bindings at all rather than about suppressing errors. I think this is probably a legitimate error case, the only binding available is not one that makes sense which sounds like a configuration issue that probably shouldn't be hidden away.

This version of `optional` is most often useful in situations where you're writing infrastructure code that loads list of some kind of interface and operates over lots of them. Think, the queue manager I mentioned above or a HTTP server loading routes. These use cases definitely care about invalid bindings and wouldn't want to silently continue in a potentially ill defined situation but don't want to fail just because the user hasn't provided any content for them yet and perhaps won't ever.
